sw/airborne/firmwares/rotorcraft/stabilization/stabilization_attitude_heli_indi.c

```diff
@@ -0,0 +1,656 @@
+  c->measurement[INDI_THRUST] = body_accel.z;
+
+  /* Get RPM measurement */
+#if STABILIZATION_ATTITUDE_HELI_INDI_USE_NOTCHFILTER
+  if (heli_indi_ctl.enable_notch) {
+    heli_indi_ctl.motor_rpm = rpm_sensor_get_rpm();
+  } else {
+    heli_indi_ctl.motor_rpm = 0;
+  }
+#endif
+
+  /* Apply actuator dynamics model to previously commanded values
+   * input  = actuator command in previous cycle
+   * output = actual actuator position right now
+   */
+  c->apply_actuator_models(c->actuator_out, c->command_out[__k - 1]);
+
+  /* Apply a set of filters, both to the actuator and the measurement */
+  c->apply_actuator_filters[0](c->filtered_actuator[0], c->actuator_out);
+  c->apply_measurement_filters[0](c->filtered_measurement[0], c->measurement);
+  for (uint8_t i = 1; i < INDI_NR_FILTERS; i++) {
+    c->apply_actuator_filters[i](c->filtered_actuator[i], c->filtered_actuator[i - 1]);
+    c->apply_measurement_filters[i](c->filtered_measurement[i], c->filtered_measurement[i - 1]);
+  }
+
+  /* RADIO throttle stick value, for 4dof mode */
+  //int32_t accel_z_sp = (-1)*3*((guidance_v_rc_delta_t - MAX_PPRZ/2) << INT32_ACCEL_FRAC) / (MAX_PPRZ/2);
+  //accel_z_sp = ((accel_z_sp << INT32_TRIG_FRAC) / guidance_v_thrust_coeff);
+
+  /* Transform yaw into a delta yaw while keeping filtered yawrate (kinda hacky) */
+  int32_t filtered_measurement_vector[INDI_DOF];
+  indi_copy_vect(filtered_measurement_vector, c->filtered_measurement[INDI_NR_FILTERS - 1]);
+  static int32_t previous_filt_yawrate = 0;
+  filtered_measurement_vector[INDI_YAW] = 512 * (c->filtered_measurement[INDI_NR_FILTERS - 1][INDI_YAW] -
+                                          previous_filt_yawrate);  // = approximately yaw acceleration error
+  previous_filt_yawrate = c->filtered_measurement[INDI_NR_FILTERS - 1][INDI_YAW];
+
+  /* Obtain virtual control input with P controller on pitch and roll */
+  int32_t roll_virtual_control  = (heli_indi_gains.roll_p * att_err.qx)  / 4;
+  int32_t pitch_virtual_control = (heli_indi_gains.pitch_p * att_err.qy) / 4;
+
+  /* Yaw with cascaded PD-controller to generate virtual control */
+  int32_t yaw_rate_reference = (heli_indi_gains.yaw_p * att_err.qz / 8);
+  int32_t yaw_virtual_control = heli_indi_gains.yaw_d * (yaw_rate_reference - body_rate->r);
+
+  /* Set INDI references */
+  c->reference[INDI_ROLL]   = roll_virtual_control;
+  c->reference[INDI_PITCH]  = pitch_virtual_control;
+  c->reference[INDI_YAW]    = yaw_virtual_control;
+  //c->reference[INDI_THRUST] = accel_z_sp;
+
+  /* Subtract (filtered) measurement from reference to get the error */
+  indi_subtract_vect(c->error, c->reference, filtered_measurement_vector);
+
+  /* Multiply error with inverse of actuator effectiveness, to get delta u (required increment in input) */
+  indi_matrix_multiply_vector(c->du, c->invG, c->error);
+
+  /* Bitshift back */
+  c->du[INDI_ROLL]  >>= 16;
+  c->du[INDI_PITCH] >>= 16;
+  c->du[INDI_YAW]   >>= 16;
+  c->du[INDI_THRUST] >>= 16;
+
```

Comment:
bitshift back from what? Where does the 16 come from? If it is a fixed bit depth can you add a define which describes the value like INT32_QUAT_FRAC